File: c/gpt/p21_fc.c
/**
 * p21_fc.c - CLI add 1000 with overflow check (C)
 *
 * Reads one numeric command-line argument, parses it robustly, checks for
 * overflow before adding 1000, prints the result on success, returns 0, and
 * exits non-zero on any validation failure.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>

static const long long ADDEND = 1000;

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Error: missing numeric argument.\n");
        return 1;
    }

    const char *arg = argv[1];
    char *endptr = NULL;
    errno = 0;
    long long value = strtoll(arg, &endptr, 10);

    if (errno != 0) {
        fprintf(stderr, "Error: parsing failed.\n");
        return 1;
    }
    if (endptr == arg || *endptr != '\0') {
        fprintf(stderr, "Error: argument is not a clean integer.\n");
        return 1;
    }

    if (value > LLONG_MAX - ADDEND) {
        fprintf(stderr, "Error: addition would overflow.\n");
        return 1;
    }
    if (value < LLONG_MIN + ADDEND) {
        fprintf(stderr, "Error: addition would underflow.\n");
        return 1;
    }

    long long result = value + ADDEND;
    printf("%lld\n", result);
    return 0;
}

File: python/gpt/p21_fc.py
"""
p21_fc.py - CLI add 1000 with overflow-like check (Python)

Reads one numeric command-line argument, parses it, checks that adding 1000
stays within a 64-bit signed integer range, prints the result on success,
and exits non-zero on validation failure. On success exits 0.
"""

from __future__ import annotations

import sys

ADDEND = 1000
INT64_MAX = 2**63 - 1
INT64_MIN = -2**63


def main() -> None:
    if len(sys.argv) < 2:
        print("Error: missing numeric argument.", file=sys.stderr)
        sys.exit(1)

    arg = sys.argv[1]
    try:
        value = int(arg, 10)
    except ValueError:
        print("Error: argument is not a valid integer.", file=sys.stderr)
        sys.exit(1)

    if value > INT64_MAX - ADDEND or value < INT64_MIN + ADDEND:
        print("Error: addition would overflow 64-bit range.", file=sys.stderr)
        sys.exit(1)

    result = value + ADDEND
    print(result)
    sys.exit(0)


if __name__ == "__main__":
    main()

